rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+3. ORANGE: At this point you only need to add three SNARKs per block: one
+   coinbase, one fee transfer, and one merge. To clear the last bit we have to
+   abandon using the parallel scan entirely. You can't empty it if you're adding
+   a coinbase to it every block. In this phase, coinbases and fee transfers can
+   no longer be added to the scan state, and we add a single extra ledger proof,
+   built sequentially, to the block data. This is a proof of the ledger state
+   transition to be applied after everything in the scan state. The block
+   producers themselves will need to create these. Block creation is otherwise
+   normal, but any necessary coinbases and fee transfers are merged into the
+   extra proof. Note this means that building these SNARKs must be possible in
+   less than one slot's time. It doesn't, however, have to be efficient to do
+   so, since this is a rare and specially incentivized situation.
+
+   For concreteness: suppose it's the first block of the orange phase. Let S_n
+   be the state hash after all transactions in the scan state are applied. When
+   Alice creates a block, she buys three SNARKs, and pays herself with a
+   coinbase. She builds base proofs for the necessary coinbase and fee transfer
+   transactions, and then a merge proof of those. The statement is S_n -> S_n+2.
+   In the following block(s), Bob only needs to buy one SNARK bundle to continue
+   draining the scan state, since he's not adding transactions to it anymore. He
+   creates a coinbase to pay himself, and it has room for the single fee
+   transfer he needs. He builds a base proof of that with statement S_n+2 ->
+   S_n+3, and a merge proof using the previous one with statement S_n -> S_n+3.
+   This process continues until the scan state is empty and a snarked ledger
+   proof up to S_n is emitted. The final block of the orange phase applies the
+   sequentially built snark, giving us a valid era n block with a ledger that
+   includes all transactions up to now.
+4. GREEN: we may now create an era n+1 block, with an empty (new-style) scan
+   state, starting from the correct state hash. Coinbase amounts go back to
+   normal.
+
+#### Checking blockchain SNARKs from prior eras
+
+The part of the blockchain SNARK that verifies the previous blockchain SNARK
+will choose which verification key to use depending on the global slot of the
+previous block, along with whether era end finalization has finished. This will
+require adding fields to `Blockchain_state` and `Consensus_state`. A circuit for
+era n only references the verification keys for era n and era n-1, it can't
+check older SNARKs.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This will be a lot of work to implement, and complicates many pieces of Coda.
+We'll have to write some serialization logic manually, type hackery is always a
+mixed blessing, we'll add another way to add ledger proofs, and in general it
```

Comment:
I think the type hackery here is warranted to avoid spaghetti conditionals that are hard to trace and verify (with the types we have spaghetti conditionals that are statically checked :shrug:)